source/extensions/filters/http/api_key_auth/api_key_auth.h

```diff
@@ -0,0 +1,201 @@
+    credentials_.reserve(proto_config.credentials().size());
+
+    for (const auto& credential : proto_config.credentials()) {
+      if (credentials_.contains(credential.key())) {
+        throw EnvoyException("Duplicate API key.");
+      }
+      credentials_[credential.key()] = credential.client();
+    }
+  }
+
+  /**
+   * To get the optional reference of the key sources.
+   */
+  OptRef<const KeySources> keySources() const {
+    return !key_sources_.empty() ? makeOptRef(key_sources_) : OptRef<const KeySources>{};
+  }
+
+  /**
+   * To get the optional reference of the credentials.
+   */
+  OptRef<const Credentials> credentials() const {
+    return !credentials_.empty() ? makeOptRef<const Credentials>(credentials_)
+                                 : OptRef<const Credentials>{};
+  }
+
+private:
+  const KeySources key_sources_;
+  Credentials credentials_;
+};
+
+class RouteConfig : public Router::RouteSpecificFilterConfig {
+public:
+  RouteConfig(const ApiKeyAuthPerRouteProto& proto_config);
+
+  /**
+   * To get the optional reference of the credentials. If this returns an valid reference, then
+   * the credentials will override the default credentials.
+   */
+  OptRef<const Credentials> credentials() const { return override_config_.credentials(); }
+
+  /**
+   * To get the optional reference of the key sources. If this returns an valid reference, then
+   * the key sources will override the default key sources.
+   */
+  OptRef<const KeySources> keySources() const { return override_config_.keySources(); }
+
+  /**
+   * To check if the client is allowed.
+   * @param client the client ID to check.
+   * @return true if the client is allowed, otherwise false.
+   */
+  bool allowClient(absl::string_view client) const {
+    return allowed_clients_.empty() || allowed_clients_.contains(client);
+  }
+
+private:
+  ApiKeyAuthConfig override_config_;
+  absl::flat_hash_set<std::string> allowed_clients_;
+};
+
+struct AuthResult {
+  bool authenticated{};
+  bool authorized{};
+  absl::string_view response_code_details{};
+};
+
+class FilterConfig {
+public:
+  FilterConfig(const ApiKeyAuthProto& proto_config, Stats::Scope& scope,
+               const std::string& stats_prefix);
+
+  /**
+   * To get the optional reference of the default credentials.
+   */
+  OptRef<const Credentials> credentials() const { return default_config_.credentials(); }
+
+  /**
+   * To get the optional reference of the default key sources.
+   */
+  OptRef<const KeySources> keySources() const { return default_config_.keySources(); }
+
+  /**
+   * To get the stats of the filter.
+   */
+  ApiKeyAuthStats& stats() { return stats_; }
+
+private:
+  static ApiKeyAuthStats generateStats(Stats::Scope& scope, const std::string& prefix) {
+    return ApiKeyAuthStats{ALL_API_KEY_AUTH_STATS(POOL_COUNTER_PREFIX(scope, prefix))};
+  }
+
+  ApiKeyAuthConfig default_config_;
```

Comment:
I suggest being strict about const correctness of the configuration types as they are shared across multiple threads.